File: ScorchedServer/ScorchedServer/Models/Rectangle.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

namespace ScorchedServer.Models
{
  public class Rectangle
  {
    public Vector LeftTop { get; private set; }
    public Vector RightTop { get; private set; }
    public Vector LeftBottom { get; private set; }
    public Vector RightBottom { get; private set; }

    public double Width { get; private set; }
    public double Height { get; private set; }
    public Vector Center { get; private set; }

    public Vector CenterTop { get; private set; }
    public Vector CenterLeft { get; private set; }
    public Vector CenterRight { get; private set; }
    public Vector CenterBottom { get; private set; }

    private Rectangle(double top, double left, double right, double bottom)
    {
      LeftTop = Vector.FromCart(left, top);
      RightTop = Vector.FromCart(right, top);
      LeftBottom = Vector.FromCart(left, bottom);
      RightBottom = Vector.FromCart(right, bottom);

      Width = right - left;
      Height = top - bottom;
      Center = Vector.FromCart(left + Width / 2, bottom + Height / 2);

      CenterTop = Vector.FromCart(Center.X, top);
      CenterLeft = Vector.FromCart(left, Center.Y);
      CenterRight = Vector.FromCart(right, Center.Y);
      CenterBottom = Vector.FromCart(Center.X, bottom);
    }

    public static Rectangle FromCenter(Vector center, double width, double height)
    {
      var top    = center.Y + height / 2;
      var left   = center.X - width / 2;
      var right  = center.X + width / 2;
      var bottom = center.Y - height / 2;

      return new Rectangle(top, left, right, bottom);
    }

    public static Rectangle FromPoints(double top, double left, double right, double bottom)
    {
      return new Rectangle(top, left, right, bottom);
    }

    public static Rectangle FromVectors(Vector v1, Vector v2)
    {
      var top    = v1.Y > v2.Y ? v1.Y : v2.Y;
      var left   = v1.X < v2.X ? v1.X : v2.X;
      var right  = v1.X > v2.X ? v1.X : v2.X;
      var bottom = v1.Y < v2.Y ? v1.Y : v2.Y;

      return new Rectangle(top, left, right, bottom);
    }
  }
}

File: src/ScorchedServer/play/rectangle.js
// Immutable
function Rectangle(top, left, right, bottom)
{
  this.top         = top;
  this.left        = left;
  this.right       = right;
  this.bottom      = bottom;
  this.leftTop     = Vector.fromCart(left, top);
  this.rightTop    = Vector.fromCart(right, top);
  this.leftBottom  = Vector.fromCart(left, bottom);
  this.rightBottom = Vector.fromCart(right, bottom);

  this.width  = this.right - this.left;
  this.height = this.top - this.bottom;
  this.center = Vector.fromCart(this.left + this.width / 2, this.bottom + this.height /2);

  this.centerTop    = Vector.fromCart(this.center.x, top);
  this.centerLeft   = Vector.fromCart(left, this.center.y);
  this.centerRight  = Vector.fromCart(right, this.center.y);
  this.centerBottom = Vector.fromCart(this.center.x, bottom);

  this.rotate = function(angle, origin)
  {
    var v1 = this.leftTop.rotate(angle, origin);
    var v2 = this.rightBottom.rotate(angle, origin);

    return Rectangle.fromVectors(v1, v2);
  }

  this.toString = function ()
  {
    return "Rectangle { top: " + this.top + ", left: " + this.left +  ", right: " + this.right +  ", bottom: " + this.bottom + "}";
  }
}

Rectangle.fromPoints = function (top, left, right, bottom)
{
  return new Rectangle(top, left, right, bottom);
}

Rectangle.fromCenter = function (center, width, height)
{
  var top    = center.y + height / 2;
  var left   = center.x - width / 2;
  var right  = center.x + width / 2;
  var bottom = center.y - height / 2;

  return new Rectangle(top, left, right, bottom);
}

Rectangle.fromVectors = function (v1, v2)
{
  var top    = v1.y > v2.y ? v1.y : v2.y;
  var left   = v1.x < v2.x ? v1.x : v2.x;
  var right  = v1.x > v2.x ? v1.x : v2.x;
  var bottom = v1.y < v2.y ? v1.y : v2.y;

  return new Rectangle(top, left, right, bottom);
}

